File: Review-15-Python-Strategy-Pattern/Example-5/tictactoe/referee.py
import itertools
from typing import Optional

from .board import VALID_SYMBOLS, Board


def _all_three_match(triple: list[str]) -> bool:
    """
    Check for three matching symbols in the Pair-Triple.

    Args:
        triple: set of three cells to check

    Returns:
        True if all three pairs contain the same symbol
    """

    first_symbol = triple[0]
    return all(symbol == first_symbol for symbol in triple)


class Referee:
    """
    The Meta-player that checks game status
    e.g., checks for wins, who won, if there is
    a stalemate.

    It is an implementation detail that is not
    exposed to the outside world.
    """

    def __init__(self, board: Board):
        """
        Create the referee and allow access
        to a board through a constant reference variable.

        Args:
            board: the board to monitor
        """

        self._board_ref = board

    def check_for_win(self) -> Optional[str]:
        """
        Check for a win condition.

        Returns:
            "X" if player1, or "O" if player2 won. None indicates neither
            player won.
        """

        triples = itertools.chain(
            self._board_ref.rows(),
            self._board_ref.columns(),
            self._board_ref.diagonals(),
        )

        # if they match, grab the 'X' or 'O'
        if matching_triple := next(filter(_all_three_match, triples), None):
            return matching_triple[0]

        return None

    def selected_cell_is_empty(self, move: int) -> bool:
        """
        Determine whether a cell in the board has been selected
        by a player.

        Args:
            move: player candidate move

        Returns:
            True if the cell is currently empty
        """

        return self._board_ref.get_cell(move) not in VALID_SYMBOLS

File: 15-Java-Strategy-Pattern/Example-3/src/main/java/edu/odu/cs/cs330/examples/Referee.java
package edu.odu.cs.cs330.examples;

import java.util.stream.Stream;
import java.util.Optional;

/**
 * The Meta-player that checks game status
 * e.g., checks for wins, who won, if there is
 * a stalemate.
 *
 * It is an implementation detail that is not
 * exposed to the outside world.
 */
public class Referee {
    /**
     * Reference to the gameboard.
     */
    private final Board boardRef;

    /**
     * This constructor must never be called.
     */
    private Referee()
    {
        boardRef = null;
    }

    /**
     * Create the referee and allow access
     * to a board through a constant reference variable.
     *
     * @param b the board to monitor
     */
    public Referee(final Board b)
    {
        this.boardRef = b;
    }

    /**
     * Utility function to check whether a character is 'X' or 'O' (i.e., a
     * player symbol).
     */
    private static boolean isPossiblePlayerSymbol(final char theChar)
    {
        return theChar == 'X' || theChar == 'O';
    }

    /**
     * Check for a win condition.
     *
     * @return 'X' if player1, or 'O' if player2 won.
     *     Any other character indicates neither player won
     */
    public char checkForWin()
    {
        Optional<Board.Pair[]> match = Stream.of(
                // Horizontal
                boardRef.get3Cells(1, 2, 3),
                boardRef.get3Cells(4, 5, 6),
                boardRef.get3Cells(7, 8, 9),
                // Vertical
                boardRef.get3Cells(1, 4, 7),
                boardRef.get3Cells(2, 5, 8),
                boardRef.get3Cells(3, 6, 9),
                // Diagonal
                boardRef.get3Cells(1, 5, 9),
                boardRef.get3Cells(7, 5, 3)
            )
            .filter(triple -> allThreeMatch(triple))
            .findFirst();

        if (match.isPresent()) {
            return match.get()[0].second;
        }

        return '?';
    }

    /**
     * Determine whether a cell in the board has been selected
     * by a player.
     *
     * @param move player candidate move
     *
     * @return true if the cell is currently empty
     */
    public boolean selectedCellIsEmpty(int move)
    {
        return boardRef.getCell(move) != 'X' && boardRef.getCell(move) != 'O';
    }

    /**
     * Check for three matching symbols in the Pair-Triple.
     *
     * @param triple set of three cells to check
     *
     * @return true if all three pairs contain the same symbol
     */
    private boolean allThreeMatch(Board.Pair[] triple)
    {
        char firstVal = triple[0].second; // std::pair first and second

        int numMatches = 0;

        for (int i = 0; i < triple.length; i++) {
            if (firstVal == triple[i].second) {
                numMatches++;
            }
        }

        return (numMatches == 3);
    }
}
